File: sdks/python/src/echo_agent_sdk/catalog.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any


@dataclass(frozen=True, slots=True)
class ResolvedOperation:
    operation: str
    family: str
    method: str
    signature: str


@dataclass(frozen=True, slots=True)
class CatalogFamily:
    """A deterministic family projection of the canonical operation catalog."""

    family: str
    methods: tuple[str, ...]
    operations: tuple[ResolvedOperation, ...]


class FacadeCatalog:
    def __init__(self, path: str | Path | None = None) -> None:
        resolved = (
            Path(path)
            if path is not None
            else Path(__file__).resolve().parents[3]
            / "shared"
            / "facade-operation-catalog.json"
        )
        self.document: dict[str, Any] = json.loads(resolved.read_text(encoding="utf-8"))
        digest_path = resolved.with_name("contract-digests.json")
        digests = json.loads(digest_path.read_text(encoding="utf-8"))
        self.contract_digest = str(digests["contract_digest"])
        self.source_contract_digest = str(digests["source_contract_digest"])
        routes = self.document.get("routes", [])
        self._operations = {
            route["operation"]: route for route in routes if route.get("operation")
        }
        self._family_routes: dict[str, list[dict[str, Any]]] = {}
        for route in routes:
            method = route.get("method")
            if route.get("operation") is None and method:
                self._family_routes.setdefault(str(method), []).append(route)

        operation_entries: list[ResolvedOperation] = []
        for route in routes:
            operation = route.get("operation")
            if operation:
                operation_entries.append(self._resolved_route(route, str(operation)))
            method = route.get("method")
            if not method:
                continue
            for item in route.get("operation_signatures", []):
                nested_operation = item.get("operation")
                if nested_operation:
                    operation_entries.append(
                        self._resolved_family_operation(
                            route, item, str(method), str(nested_operation)
                        )
                    )

        operation_entries.sort(
            key=lambda item: (item.operation, item.family, item.method, item.signature)
        )
        identities = [
            (item.operation, item.family, item.method, item.signature)
            for item in operation_entries
        ]
        if len(identities) != len(set(identities)):
            raise ValueError(
                "canonical catalog contains duplicate operation identities"
            )
        self._operation_entries = tuple(operation_entries)

        grouped_operations: dict[str, list[ResolvedOperation]] = {}
        for item in self._operation_entries:
            grouped_operations.setdefault(item.family, []).append(item)
        family_entries: list[CatalogFamily] = []
        family_names = {
            str(item["family"])
            for item in self.document.get("families", [])
            if item.get("family")
        }
        family_names.update(grouped_operations)
        for family_name in sorted(family_names):
            metadata = next(
                (
                    item
                    for item in self.document.get("families", [])
                    if item.get("family") == family_name
                ),
                {},
            )
            methods = {str(method) for method in metadata.get("methods", []) if method}
            methods.update(
                item.method for item in grouped_operations.get(family_name, [])
            )
            family_entries.append(
                CatalogFamily(
                    family=family_name,
                    methods=tuple(sorted(methods)),
                    operations=tuple(grouped_operations.get(family_name, [])),
                )
            )
        self._family_entries = tuple(family_entries)

    @staticmethod
    def _signature(route: dict[str, Any], context: str) -> str:
        digests = route.get("signature_digests", [])
        if not digests:
            raise ValueError(f"canonical catalog route has no signature: {context}")
        return str(digests[0])

    @classmethod
    def _resolved_route(
        cls, route: dict[str, Any], operation: str
    ) -> ResolvedOperation:
        family = route.get("family")
        method = route.get("method")
        if not family or not method:
            raise ValueError(
                f"canonical source route is missing family/method: {operation}"
            )
        return ResolvedOperation(
            operation,
            str(family),
            str(method),
            cls._signature(route, operation),
        )

    @classmethod
    def _resolved_family_operation(
        cls,
        route: dict[str, Any],
        item: dict[str, Any],
        method: str,
        operation: str,
    ) -> ResolvedOperation:
        family = route.get("family")
        digests = item.get("signature_digests", [])
        if not family or not digests:
            raise ValueError(
                f"canonical family route is missing family/signature: {method}/{operation}"
            )
        return ResolvedOperation(operation, str(family), method, str(digests[0]))

    def operations(self) -> tuple[ResolvedOperation, ...]:
        """Return every source and family operation in canonical order."""

        return self._operation_entries

    def source_operations(self) -> tuple[ResolvedOperation, ...]:
        """Return source-operation routes in canonical order."""

        return tuple(
            item
            for item in self._operation_entries
            if item.family == "source_operation"
        )

    def family_operations(self) -> tuple[ResolvedOperation, ...]:
        """Return nested family-operation routes in canonical order."""

        return tuple(
            item
            for item in self._operation_entries
            if item.family != "source_operation"
        )

    def families(self) -> tuple[CatalogFamily, ...]:
        """Return all canonical families, including families with no operations."""

        return self._family_entries

    def signature(self, operation: str) -> str:
        route = self._operations.get(operation)
        digests = route.get("signature_digests", []) if route else []
        if not digests:
            raise KeyError(f"operation is absent from canonical catalog: {operation}")
        return str(digests[0])

    def family_signature(self, method: str, operation: str) -> str:
        for route in self._family_routes.get(method, []):
            for item in route.get("operation_signatures", []):
                if item.get("operation") == operation and item.get("signature_digests"):
                    return str(item["signature_digests"][0])
        raise KeyError(
            f"family operation is absent from canonical catalog: {method}/{operation}"
        )

    def resolve(self, operation: str) -> ResolvedOperation:
        direct = self._operations.get(operation)
        if direct and direct.get("method") and direct.get("signature_digests"):
            return ResolvedOperation(
                operation,
                str(direct["family"]),
                str(direct["method"]),
                str(direct["signature_digests"][0]),
            )
        for method, routes in self._family_routes.items():
            for route in routes:
                for item in route.get("operation_signatures", []):
                    if item.get("operation") == operation and item.get(
                        "signature_digests"
                    ):
                        return ResolvedOperation(
                            operation,
                            str(route["family"]),
                            method,
                            str(item["signature_digests"][0]),
                        )
        raise KeyError(f"operation is absent from the canonical catalog: {operation}")

File: sdks/typescript/src/catalog.ts
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

export interface CatalogRoute {
  readonly family: string;
  readonly method: string | null;
  readonly operation: string | null;
  readonly handler_operation?: string | null;
  readonly route: string;
  readonly signature_digests?: readonly string[];
  readonly operation_signatures?: readonly {
    readonly operation: string;
    readonly signature_digests: readonly string[];
  }[];
  readonly required_feature?: string | null;
  readonly required_features?: readonly string[];
  readonly feature_semantics?: string;
}

export interface ResolvedOperation {
  readonly operation: string;
  readonly family: string;
  readonly method: string;
  readonly signature: string;
}

export interface CatalogOperation {
  /** Stable identity scoped by the catalog family. */
  readonly identity: string;
  readonly operation: string;
  readonly family: string;
  readonly method: string;
  readonly route: string;
  readonly signature: string;
  readonly signature_digests: readonly string[];
  readonly required_feature: string | null;
  readonly required_features: readonly string[];
  readonly feature_semantics: string;
}

export interface CatalogFamily {
  readonly family: string;
  readonly methods: readonly string[];
  readonly operations: readonly CatalogOperation[];
  readonly required_feature: string | null;
}

export interface CatalogDocument {
  readonly schema_version: number;
  readonly extension_protocol_version: number;
  readonly total_items: number;
  readonly routes: readonly CatalogRoute[];
  readonly families: readonly {
    readonly family: string;
    readonly methods: readonly string[];
    readonly operations: readonly string[];
    readonly required_feature?: string | null;
  }[];
}

export class FacadeCatalog {
  readonly document: CatalogDocument;
  readonly contractDigest: string;
  readonly sourceContractDigest: string;
  private readonly operationRoutes: ReadonlyMap<string, CatalogRoute>;
  private readonly familyRoutes: ReadonlyMap<string, CatalogRoute>;
  private readonly operationList: readonly CatalogOperation[];
  private readonly familyList: readonly CatalogFamily[];

  constructor(path?: string) {
    const resolved = path ?? fileURLToPath(new URL("../../shared/facade-operation-catalog.json", import.meta.url));
    this.document = JSON.parse(readFileSync(resolved, "utf8")) as CatalogDocument;
    const digestPath = path
      ? join(dirname(resolved), "contract-digests.json")
      : fileURLToPath(new URL("../../shared/contract-digests.json", import.meta.url));
    const digests = JSON.parse(readFileSync(digestPath, "utf8")) as {
      contract_digest: string;
      source_contract_digest: string;
    };
    this.contractDigest = digests.contract_digest;
    this.sourceContractDigest = digests.source_contract_digest;
    this.operationRoutes = new Map(
      this.document.routes
        .filter((route) => route.operation !== null)
        .map((route) => [route.operation as string, route]),
    );
    this.familyRoutes = new Map(
      this.document.routes
        .filter((route) => route.operation === null)
        .filter((route) => route.method !== null)
        .map((route) => [route.method as string, route]),
    );

    const operations = this.document.routes.flatMap((route) => {
      const direct = route.operation && route.method
        ? [this.toCatalogOperation(route, route.operation, route.signature_digests ?? [])]
        : [];
      const family = route.method
        ? (route.operation_signatures ?? []).map((entry) =>
            this.toCatalogOperation(route, entry.operation, entry.signature_digests),
          )
        : [];
      return [...direct, ...family];
    });
    operations.sort((left, right) => left.identity.localeCompare(right.identity));
    this.operationList = operations;

    const operationsByFamily = new Map<string, CatalogOperation[]>();
    for (const operation of operations) {
      const family = operationsByFamily.get(operation.family) ?? [];
      family.push(operation);
      operationsByFamily.set(operation.family, family);
    }
    this.familyList = this.document.families
      .map((family) => ({
        family: family.family,
        methods: [...new Set([
          ...family.methods,
          ...(operationsByFamily.get(family.family) ?? []).map((operation) => operation.method),
        ])].sort(),
        operations: operationsByFamily.get(family.family) ?? [],
        required_feature: family.required_feature ?? null,
      }))
      .sort((left, right) => left.family.localeCompare(right.family));
  }

  operations(): readonly CatalogOperation[] {
    return this.operationList;
  }

  families(): readonly CatalogFamily[] {
    return this.familyList;
  }

  signature(operation: string): string {
    const digests = this.operationRoutes.get(operation)?.signature_digests;
    const digest = digests?.[0];
    if (!digest) throw new Error(`operation is absent from the canonical catalog: ${operation}`);
    return digest;
  }

  familySignature(method: string, operation: string): string {
    const route = this.familyRoutes.get(method);
    const digest = route?.operation_signatures
      ?.find((entry) => entry.operation === operation)
      ?.signature_digests[0];
    if (!digest) throw new Error(`family operation is absent from the canonical catalog: ${method}/${operation}`);
    return digest;
  }

  resolve(operation: string): ResolvedOperation {
    const direct = this.operationRoutes.get(operation);
    if (direct?.method && direct.signature_digests?.[0]) {
      return {
        operation,
        family: direct.family,
        method: direct.method,
        signature: direct.signature_digests[0],
      };
    }
    for (const route of this.familyRoutes.values()) {
      const entry = route.operation_signatures?.find((item) => item.operation === operation);
      if (route.method && entry?.signature_digests[0]) {
        return {
          operation,
          family: route.family,
          method: route.method,
          signature: entry.signature_digests[0],
        };
      }
    }
    throw new Error(`operation is absent from the canonical catalog: ${operation}`);
  }

  route(operation: string): CatalogRoute {
    const route = this.operationRoutes.get(operation);
    if (!route) throw new Error(`operation is absent from the canonical catalog: ${operation}`);
    return route;
  }

  private toCatalogOperation(
    route: CatalogRoute,
    operation: string,
    signatureDigests: readonly string[],
  ): CatalogOperation {
    const method = route.method;
    const signature = signatureDigests[0];
    if (!method || !signature) {
      throw new Error(`canonical catalog operation is missing method or signature: ${operation}`);
    }
    return {
      identity: `${route.family}:${operation}`,
      operation,
      family: route.family,
      method,
      route: route.route,
      signature,
      signature_digests: [...signatureDigests],
      required_feature: route.required_feature ?? null,
      required_features: [...(route.required_features ?? [])],
      feature_semantics: route.feature_semantics ?? "default",
    };
  }
}
